wagtail/core/models.py

```diff
@@ -36,9 +39,117 @@
 logger = logging.getLogger('wagtail.core')
 
 PAGE_TEMPLATE_VAR = 'page'
+SITE_CACHE_SITES_KEY = 'wagtail.core.sites'
+SITE_CACHE_DEFAULT_KEY = '__default__'
 
 
 class SiteManager(models.Manager):
+
+    @classmethod
+    def get_site_cache(cls):
+        """Returns a cache instance to use for site caching operations, or
+        ``None`` if site caching isn't enabled, or is disabled due to the
+        specified cache having an incompatible backend.
+        """
+        enabled = getattr(settings, 'WAGTAIL_SITE_CACHE_ENABLED', False)
+        if not enabled:
+            return
+
+        key = getattr(settings, 'WAGTAIL_SITE_CACHE', 'default')
+        cache = caches[key]
+        if isinstance(cache, DummyCache):
+            warnings.warn(
+                "WAGTAIL_SITE_CACHE is set to '{0}', which refers to a cache "
+                "that utilises Django's DummyCache backend. Wagtail's site "
+                "caching feature is incompatible with DummyCache, and so has "
+                "been disabled. To enable it, update '{0}' to use a backend "
+                "that actually stores and returns stored values, such as "
+                "Memcached or Redis. If multiple CACHES are configured, you"
+                "can use the WAGTAIL_SITE_CACHE setting to specify a cache "
+                "other than '{0}' to use instead".format(key))
+            return
+        if not settings.DEBUG and isinstance(cache, LocMemCache):
+            warnings.warn(
+                "WAGTAIL_SITE_CACHE is set to '{0}', which refers to a cache "
+                "that utilises Django's LocMemCache backend. Wagtail's site "
+                "caching feature is incompatible with LocMemCache when running "
+                "in a production environment, and so has been disabled. To "
```

Comment:
Would it be better to more explicitly refer to the use of `settings.DEBUG` here, instead of "in a production environment"? I can think of use cases where someone might be testing with `DEBUG=False` when not in production, and might not be sure what this is referring to.

Reply:
@chosak I was hoping for feedback on this wording, and that sounds like a sensible change. I was also thinking about rewording so that  "Wagtail's site caching has been disabled" comes first?

Reply:
Personally I find it more readable this way: explaining an incompatibility and then saying, in conclusion, "and so has been disabled". But others might have different or stronger opinions.

Relatedly, can you provide a little more detail behind the rationale of preventing this cache in production? The [Django docs say](https://docs.djangoproject.com/en/2.0/topics/cache/#local-memory-caching) that the `LocMemCache` is thread-safe, although, on the other hand, is only per-process, so wouldn't be shared among e.g. multiple server threads on the same instance. But if that's the concern, isn't that also a problem when running across multiple instances -- for example, modifying a Site through the admin that hits one instance wouldn't know how to clear caches on the other instances. To be totally safe, doesn't this caching only work if there's a single site cache for all Django processes in a single Wagtail deployment?

Reply:
Yes, 'thread safety' is probably the wrong term here. What I am really trying to guard against is use of `LocMemCache` when the site is running as multiple processes, because each process would have it's own version of the cache in memory, and signals would only update caches in the processes where they were triggered. I'm open to recommendations on how to word this?

Any running processes will indeed need to share the same cache instance in order for the cached data to remain consistent between processes. This will no doubt have to be outlined in the docs.

Reply:
Right, so doesn't that just as equally apply to the file-based cache, for Wagtail deployments that run across multiple servers? Having a `LocMemCache`-specific callout/check might be a bit misleading if there are other caches that would also need to be excluded -- not to mention any custom caches that someone might be using.

Reply:
Oh dang, you're right. I completely overlooked that! I suppose it's a bit of a grey area as to how far we go in terms of accounting for things like this. One option would be to not to give any special treatment to specific backends at all and just explain as well as possible in the docs. Personally, I'd prefer to treat FileBasedCache in the same way as LocMemCache and leave it at that - that's at least Django's official backends covered. I'm open to other ideas though?

Reply:
My preference would probably be to remove special treatment of any backends, and leave it to the documentation to warn users about sensible cache configuration. Pinging @gasman @lb- for their opinions.

Speaking of: is `DummyCache` really "incompatible" or does it just work as a no-op in this case that ends up being less efficient than the default behavior? If so, I wonder if that check really adds any value, as it'd work the same way as any other time that cache backend is used.

Reply:
`DummyCache` was entirely incompatible (which was one of my reasons for wanting to refactor, although I don't think I mentioned it) - But now, it is just as you say - it will work fine, but it will be a little less efficient.

I'm almost certain that @gasman and @lb's thoughts will reflect yours, as realistically, there are more arguments for ditching the special case treatment than there is for keeping it - and I guess I'm fine with that really, so long as the documentation is right.

I'll work on this later :)